core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -334,6 +334,19 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
   final def ensuring(finalizer: IO[Nothing, Unit]): IO[E, A] =
     new IO.Ensuring(self, finalizer)
 
+  /**
+   * Executes the action on the specified `ExecutionContext` and then shifts back
+   * to the default one.
+   */
+  final def on(ec: ExecutionContext): IO[E, A] =
+    IO.shift(ec) *> self <* IO.shift
```

Comment:
Oops, I missed this. It should definitely be done using `bracket` (if you want it to definitely return).

Also this version won't work. Exactly. It will never hop back. You need to create an `async` boundary that waits for the (forked) shifted IO to complete, then when it resumes, it will resume on its `rts`, thus returning to the original thread pool.

Reply:
shift submits a new task to the RTS ec, so, in my understanding, it should return to the RTS after `self`. (Also, finalizers currently run on separate fibers, shifts there wouldn't impact the parent AFAIK)

Reply:
```
val ec = ExecutionContext.fromExecutorService(Executors.newCachedThreadPool())

def printThread: IO[Nothing, Unit] =
  IO.sync(println(Thread.currentThread()))

new RTS {}.unsafeRun {
  for {
    _ <- printThread
    _ <- printThread.on(ec)
    _ <- printThread
  } yield ()
}
```

Output:
```
Thread[main,5,main]
Thread[pool-1-thread-1,5,main]
Thread[zio-1-1468177767,5,zio]
```

> Also, finalizers currently run on separate fibers, shifts there wouldn't impact the parent AFAIK

Is this true? This definitely impacts `FiberLocal.locally` on cancellation. How come it forks for each finalizer?

Reply:
`.run` is used to collect defects in finalizer to propagate them to `ExitResult`, #313 might address the efficiency

Reply:
Great, I think that will definitely help out, thanks!

On the other hand, there's an impressive set of things that will non-transparently fork and break `FiberLocal`, like `bracket0`. I will go through and create a list of them.

Reply:
That would be helpful. Then I think we have to eliminate as many of them as possible, or possibly, extend the `FiberLocal` model to handle forking, e.g. you automatically share the `FiberLocal` values of your parent fiber, in a copy-on-write model that ensures children only mutate `FiberLocal` locally.